x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -0,0 +1,414 @@
+                    logger.error(() ->
+                        new ParameterizedMessage("failed to fetch range [{}-{}] of [{}]", range.v1(), range.v2(), path.getFileName()), e);
+                    future.completeExceptionally(e);
+                }
+            });
+
+            if (gaps.isEmpty() == false) {
+                fetchMissingRanges(gaps);
+            }
+
+            try {
+                return future.get();
+            } catch (InterruptedException| ExecutionException e) {
+                if (e instanceof InterruptedException) {
+                    Thread.currentThread().interrupt();
+                }
+                throw new IOException("Failed to fetch range [{}-{}] of [{}]: " + toString(), e);
+            }
+        }
+
+        /**
+         * Fetches all missing ranges
+         *
+         * @param gaps the missing ranges to fetch
+         */
+        private void fetchMissingRanges(final List<SparseFileTracker.Gap> gaps) {
+            assert gaps.isEmpty() || gaps.size() == 1;
+            for (SparseFileTracker.Gap gap : gaps) {
+                threadPool.executor(ThreadPool.Names.GENERIC).execute(new AbstractRunnable() {
+                    @Override
+                    public void onFailure(Exception e) {
+                        gap.onFailure(e);
+                    }
+
+                    @Override
+                    protected void doRun() throws Exception {
+                        fetchAndWriteRange(gap.start, gap.end);
+                        gap.onResponse(null);
+                    }
+                });
+            }
+        }
+
+        /**
+         * Fetches a missing range from the cache entry's source and writes it into the sparse cached file.
+         * <p>
+         * Even though {@link SparseFileTracker} prevents the same range to be concurrently fetched and written to disk, this method
+         * is synchronized for extra safety.
+         *
+         * @param start the range start to fetch
+         * @param end   the range end to fetch
+         * @throws IOException if something went wrong
+         */
+        private synchronized void fetchAndWriteRange(final long start, final long end) throws IOException {
+            logger.trace(() -> new ParameterizedMessage("fetching missing range [{}-{}] from source [{}]", start, end, source));
+
+            final int copyBufferSize = 8192;
+            byte[] copyBuffer = new byte[copyBufferSize];
+
+            try (IndexInput input = source.get()) {
+                if (start > 0) {
+                    input.seek(start);
+                }
+                long remaining = end - start;
+                long pos = start;
+                while (remaining > 0) {
+                    int len = (remaining < copyBufferSize) ? (int) remaining : copyBufferSize;
+                    logger.trace(() -> new ParameterizedMessage("reading {} bytes from [{}]", len, input));
+                    input.readBytes(copyBuffer, 0, len);
```

Comment:
I agree this is defeating the purpose of requesting large chunks of data from the underlying `SearchableSnapshotIndexInput`, which is today implemented as a `BufferedIndexInput`. With the current implementation only chunks of <<BufferedIndexInput's internal buffer size>> are requested (ie 1024).

I think that the right think to do is to have `SearchableSnapshotIndexInput` extends `IndexInput` directly and maintains an internal `InputStream` which is opened from the `BlobContainer`'s `readBlob(name, position, len)` method where `len` is 32Mb. 

The `SearchableSnapshotIndexInput` should take care of opening the `InputStream` from the required position (by using range bytes S3 requests), closing it and reopening it on `IndexInput`'s seekings. It should also take care of draining/aborting the InputStream in case not all bytes were read.